### ISSUE TITLE
Improving "action.prepare" performance

### ISSUE THREAD

**Issue by chiragramani** [bazelbuild/bazel#19266]
### Description of the bug:

action.prepare step - 4.4 seconds
ActionContinuation.execute - 10.4 seconds

Since all of this is in critical path where every second savings matter, we would like to explore if there are opportunities for improving the performance of the `action.prepare step`.

From the [documentation here](https://github.com/bazelbuild/bazel/blob/0ffba8189364dc86f4ee31f3b6be5e3ebd5d09d7/src/main/java/com/google/devtools/build/lib/skyframe/SkyframeActionExecutor.java#L1016), 
```
// This call generally deletes any files at locations that are declared outputs of the
// action, although some actions perform additional work, while others intentionally
// keep previous outputs in place.
```
<img width="598" alt="Screenshot 2023-08-16 at 4 27 23 PM" src="https://github.com/bazelbuild/bazel/assets/11925399/247d020f-9313-4912-876c-3f639c50266b">

### Which category does this issue belong to?

Performance

### Which operating system are you running Bazel on?

macOS

### What is the output of `bazel info release`?

Bazel 6.4

**Comment by brentleyjones** [comment]
I believe most of that time is deleting previous artifacts. In the case of tree artifacts it would probably be a lot more efficient to move the directory to a temp directory and then asynchronously delete that directory (and since it's in a temp directory, if the process is cancelled, the OS will eventually clean it up).

**Comment by tjgq** [comment]
> I believe most of that time is deleting previous artifacts.

Yes, deleting large tree artifacts is known to be slow, especially on MacOS where filesystem performance leaves a lot to be desired. I've also observed large delays when deleting stale sandboxes left over by previous builds, which would likely also benefit from your proposed solution (although, to be clear, these would manifest themselves differently in a profile).

> move the directory to a temp directory and then asynchronously delete that directory (and since it's in a temp directory, if the process is cancelled, the OS will eventually clean it up).

Moving to an OS-designated temporary directory (e.g. `/tmp`) might cross filesystems and end up being just as inefficient. I think I might prefer to move them into a designated subdirectory of the output base, and have them reaped by a background thread.

One thing I want to be careful with is to wait for any pending deletions to finish before we consider the invocation finished, as otherwise they could interfere with the performance of subsequent invocations in a benchmarking context.

**Comment by brentleyjones** [comment]
> One thing I want to be careful with is to wait for any pending deletions to finish before we consider the invocation finished, as otherwise they could interfere with the performance of subsequent invocations in a benchmarking context.

Looks like Bazel already thought of all of that, but only in the context of sandbox trees:

https://github.com/bazelbuild/bazel/blob/253e9862112e7ab6bdbe5d0b3fadabb8e544315d/src/main/java/com/google/devtools/build/lib/sandbox/AsynchronousTreeDeleter.java#L31-L39

I bet that could be massaged to work for your case as well.

**Comment by tjgq** [comment]
Great spelunking!

It looks like there's both a [synchronous](https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/sandbox/SynchronousTreeDeleter.java;l=22;drc=70b43db35dddb1af05dd5646990e6de85167ab00) and an [asynchronous](https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/sandbox/AsynchronousTreeDeleter.java;l=39;drc=70b43db35dddb1af05dd5646990e6de85167ab00) TreeDeleter implementation, but the [default](https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/sandbox/SandboxModule.java;l=213;drc=79745b4db3ca8ab42451f9eb212994a340e3f7c9) is synchronous and you have to set [`--experimental_sandbox_async_tree_delete_idle_threads`](https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/sandbox/SandboxOptions.java;l=328;drc=0ffba8189364dc86f4ee31f3b6be5e3ebd5d09d7) to make it asynchronous.

@larsrc-google, do you know why asynchronous isn't the default? (I vaguely remember discussing this with you, but I don't recall the conclusion.)

**Comment by larsrc-google** [comment]
I think we discussed it in connection with sandbox cleanup for reuse, where I suspected async deletion of causing rare errors. But I don't know why it isn't turned on for non-reusing sandboxes, it was created before my time on Bazel (https://github.com/bazelbuild/bazel/commit/16af94c749b21a96585e6c3e8a3704fb850ff86d).

**Comment by erikkerber** [comment]
We've had a developer come to us with pretty bad build slowness due to the operations behind `action.prepare`. My best guess based on discussion with colleagues that its bound by disk I/O and file deletes, all on a Macbook with endpoint security software installed. It's a rules_apple/swift based iOS project.

Setting `--experimental_sandbox_async_tree_delete_idle_threads=4` hasn't improved things, so I'm moving on to telling them to drop jobs to `HOST_CPUS` down from 200. At least as a potential (?) workaround.

![Image](https://github.com/user-attachments/assets/92fa483c-5b9b-439a-8d5c-6dbdda7005c7)

**Comment by oquenchil** [comment]
As far as I can tell [action.prepare](https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/skyframe/SkyframeActionExecutor.java;drc=bf2a2d2ff37c50dcc371423bf7b1a35e74562ac9;l=1070) is Skyframe related, not sandboxing. 

I don't think asynchronous deletion is kicking in here because it's not wired see [OutputService.java](https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/vfs/OutputService.java;drc=58680f6b28b7d49ac4ae287e6d427102dddbdfdd;l=262), [BazelOutputService.java](https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/remote/BazelOutputService.java) and [AbstractAction.java](https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/actions/AbstractAction.java;drc=634f992c57406d22cf85ba06f0b668899550c20f;l=457). It's deleting files individually.

As far as I can tell nothing implements the `BulkDeleter` in Bazel but we could make `AsynchronousTreeDeleter` implement it and add it to `BazelOutputService`.

I will discuss this with the team next week.

**Comment by brentleyjones** [comment]
Bump 🙏.
